firewall/master.go

```diff
@@ -31,91 +31,170 @@ import (
+				return
+			}
+		default: // netutils.Invalid
+			conn.Deny("invalid IP") // Block Outbound / Drop Inbound
+			return
+		}
+	} else if conn.Entity.Domain != "" {
+		// DNS Query
+		// DNS is expected to resolve to LAN or Internet addresses
+		// TODO: handle domains mapped to localhost
+		if p.BlockScopeInternet() && p.BlockScopeLAN() {
+			conn.Block("Internet and LAN access blocked")
+			return
+		}
+	}
+
+	// check endpoints list
+	var result endpoints.EPResult
+	var reason string
+	if conn.Inbound {
+		result, reason = p.MatchServiceEndpoint(conn.Entity)
+	} else {
+		result, reason = p.MatchEndpoint(conn.Entity)
+	}
 	switch result {
-	case endpoints.Undeterminable:
-		comm.UpdateVerdict(network.VerdictUndeterminable)
-		return
 	case endpoints.Denied:
-		log.Infof("firewall: denying communication %s, domain is blacklisted: %s", comm, reason)
-		comm.Deny(fmt.Sprintf("domain is blacklisted: %s", reason))
+		conn.Deny("endpoint is blacklisted: " + reason) // Block Outbound / Drop Inbound
 		return
 	case endpoints.Permitted:
-		log.Infof("firewall: permitting communication %s, domain is whitelisted: %s", comm, reason)
-		comm.Accept(fmt.Sprintf("domain is whitelisted: %s", reason))
+		conn.Accept("endpoint is whitelisted: " + reason)
+		return
+	}
+	// continuing with result == NoMatch
+
+	// implicit default=block for inbound
+	if conn.Inbound {
+		conn.Drop("endpoint is not whitelisted (incoming is always default=block)")
 		return
 	}
-	// continueing with result == NoMatch
 
 	// check default action
 	if p.DefaultAction() == profile.DefaultActionPermit {
-		log.Infof("firewall: permitting communication %s, domain is not blacklisted (default=permit)", comm)
-		comm.Accept("domain is not blacklisted (default=permit)")
+		conn.Accept("endpoint is not blacklisted (default=permit)")
 		return
 	}
 
 	// check relation
 	if !p.DisableAutoPermit() {
-		if checkRelation(comm) {
+		related, reason := checkRelation(conn)
+		if related {
+			conn.Accept(reason)
 			return
 		}
 	}
 
 	// prompt
 	if p.DefaultAction() == profile.DefaultActionAsk {
-		prompt(comm, nil, nil)
+		prompt(conn, pkt)
 		return
 	}
 
 	// DefaultAction == DefaultActionBlock
-	log.Infof("firewall: denying communication %s, domain is not whitelisted (default=block)", comm)
-	comm.Deny("domain is not whitelisted (default=block)")
+	conn.Deny("endpoint is not whitelisted (default=block)")
 	return
 }
 
 // FilterDNSResponse filters a dns response according to the application profile and settings.
-func FilterDNSResponse(comm *network.Communication, q *resolver.Query, rrCache *resolver.RRCache) *resolver.RRCache { //nolint:gocognit // TODO
+func FilterDNSResponse(conn *network.Connection, q *resolver.Query, rrCache *resolver.RRCache) *resolver.RRCache { //nolint:gocognit // TODO
 	// do not modify own queries - this should not happen anyway
-	if comm.Process().Pid == os.Getpid() {
+	if conn.Process().Pid == os.Getpid() {
 		return rrCache
 	}
```

Comment:
If this is supposed to not happen we should log it!

Reply:
Currently this is called everytime. I removed the second part of the comment.